README.md
# Orengo/Baker FunFam collaboration

This repo is currently here to store data for a working collaboration between the [Baker group](https://www.bakerlab.org/)
and the [Orengo group](http://orengogroup.info/) regarding modelling protein structures for Functional Family (FunFam)
alignments in CATH.

## Datasets

### 01.test

Initial set of 5 FunFam protein sequence alignments (STOCKHOLM format) to use for general testing: 

[`dataset/01.test/01.test.sto.tgz`](https://github.com/sillitoe/orengo-baker-funfams/blob/master/dataset/01.test/01.test.sto.tgz)

Selection criteria:

* 1 EC annotation
* reasonable number of sequences
* reasonable sequence diversity
* not too high gap percentage

### 02.medium

Wider set of FunFam alignments to test different aspects of pipeline:

Filter all funfams by:

 * high dops (>70)
 * at least 5 sequences

Then include funfams with the following characteristics:

 * lots of sequences
 * not many sequences
 * with structural reps
 * without structural reps

Within all those groups, select:

 * variety of unique folds

This dataset is generated by the script [`scripts/02.medium.py`](https://github.com/sillitoe/orengo-baker-funfams/blob/master/scripts/02.medium.py) which does the following:

 * download info on all Funfams via API
 * apply selection criteria
 * download Funfam alignments via API
 * output data file with a summary of Funfam info

The data file has been committed to this repo:

[`dataset/02.medium/02.medium.all.tsv`](https://github.com/sillitoe/orengo-baker-funfams/blob/master/dataset/02.medium/02.medium.all.tsv)


Running the script:

```sh
# setup local env
python3 -m venv venv
. venv/bin/activate
pip install -r requirements.txt

# run script
$ PYTHONPATH=. python3 scripts/02.medium.py

LOADING FUNFAM DATA
-------------------
Loading cached Funfam data from 'cath.v4_2_0.funfam.json' ...
Found 68065 funfams from 2623 unique superfamilies and 1327 unique folds

APPLYING GLOBAL FILTERS
-----------------------
High sequence diversity (seed_dops_score >= 70):             19053 (from 68065)

CREATING DATASET CATEGORIES
---------------------------
Small Funfams (num_members < 50):                            2633 (from 19053)
Large Funfams (num_members >= 5000):                         1492 (from 19053)
Funfams with structure (rep_source_id="cath"):               8464 (from 19053)
Funfams without structure (rep_source_id="uniprot"):         10589 (from 19053)

WRITING FUNFAM ALIGNMENTS
-------------------------

Working on 2633 funfams with low sequences ...
----------------------------------------------
Funfams with unique topologies:                              476 (from 1327)
GET https://www.cathdb.info/version/v4_2_0/api/rest/superfamily/1.10.10.10/funfam/20119/files/seed_alignment
Saving Funfam alignment 1.10.10.10-ff-20119
[--8<--]

Working on 1492 funfams with high sequences ...
-----------------------------------------------
Funfams with unique topologies:                              178 (from 1327)
GET https://www.cathdb.info/version/v4_2_0/api/rest/superfamily/1.10.1030.10/funfam/3118/files/seed_alignment
[--8<--]

Working on 8464 funfams with structure ...
------------------------------------------
Funfams with unique topologies:                              453 (from 1327)
GET https://www.cathdb.info/version/v4_2_0/api/rest/superfamily/1.10.1000.11/funfam/894/files/seed_alignment
[--8<--]

Working on 10589 funfams with no structure ...
----------------------------------------------
Funfams with unique topologies:                              42 (from 1327)
GET https://www.cathdb.info/version/v4_2_0/api/rest/superfamily/1.10.1510.10/funfam/1175/files/seed_alignment
[--8<--]

WRITING DATASETS
----------------
Total number of funfams                                      1149 (from 68065)
Writing TSV datafile 'dataset/02.medium/02.medium.all.tsv'
DONE

```

## Notes

### Alignment format

These files should adhere to the general [Stockholm format](https://en.wikipedia.org/wiki/Stockholm_format) rules, so you should be able to use any parser to convert them to FASTA.

We found that the alignment model within [BioPython](https://biopython.org/wiki/AlignIO) didn't provide the flexibility we needed so we added an alignment module to [`cathpy`](https://github.com/UCL/cathpy) that might be useful (see [`cathpy.align`](https://cathpy.readthedocs.io/en/latest/align.html)).

Converting from STOCKHOLM to FASTA with this package:

```python
from cathpy.align import Align
aln = Align.from_stockholm('path/to/aln.sto')
aln.write_fasta('path/to/aln.fa')
```


ChangeLog
# Changelog

## v0.1.0

*Release date: 16/08/2019*

* initial dataset: 01.test

## v0.2.0

*Release date: 11/10/2019*

* dataset: 02.medium
* added script to generate dataset directly from API (and download alignments)


cathbaker/models.py
# core
import logging

LOG = logging.getLogger(__name__)

class FunfamInfo:
    """
    Stores information from Funfam API
    """

    def __init__(self, *, cath_version, name, funfam_number, superfamily_id, rep_id, rep_source_id, 
        seed_dops_score, inclusion_bitscore, inclusion_e_value,
        num_members_in_funfam, num_members_in_seed_aln, **kwargs):

        if seed_dops_score is None:
            seed_dops_score = 0
        if inclusion_bitscore is None:
            inclusion_bitscore = 0
        if inclusion_e_value is None:
            inclusion_e_value = 0
        if num_members_in_seed_aln is None:
            num_members_in_seed_aln = 0
        if num_members_in_funfam is None:
            num_members_in_funfam = 0

        self.cath_version = cath_version

        self.name = name
        self.funfam_number = int(funfam_number)
        self.superfamily_id = superfamily_id
        self.rep_id = rep_id
        self.rep_source_id = rep_source_id
        self.seed_dops_score = float(seed_dops_score)
        self.inclusion_bitscore = float(inclusion_bitscore)
        self.inclusion_e_value = float(inclusion_e_value)
        self.num_members_in_funfam = int(num_members_in_funfam)
        self.num_members_in_seed_aln = int(num_members_in_seed_aln)

    @property
    def topology_id(self):
        """Return the C.A.T code"""
        return '.'.join(self.superfamily_id.split('.')[:3])

    @property
    def arch_id(self):
        """Return the C.A code"""
        return '.'.join(self.superfamily_id.split('.')[:2])

    def __str__(self):
        return f'{self.superfamily_id}/FF/{str(self.funfam_number)}: "{self.name}" (DOPS: {str(self.seed_dops_score)})'


cathbaker/loaders.py
# core
import json
import logging
import os

# non-core
import requests
from cathpy.align import Align

# local
from cathbaker import models

LOG = logging.getLogger(__name__)

API_FUNFAM_BASE = 'https://www.cathdb.info/version/{cath_version}/api/rest/funfam'
API_FUNFAM_ALN = 'https://www.cathdb.info/version/{cath_version}/api/rest/superfamily/{superfamily_id}/funfam/{funfam_number}/files/seed_alignment'

def get_funfam_alignment(*, cath_version, superfamily_id, funfam_number, **kwargs):
    """
    Given a FunfamInfo
    """
    ff_url = API_FUNFAM_ALN.format(cath_version=cath_version, superfamily_id=superfamily_id, funfam_number=funfam_number)
    
    LOG.info("GET %s", ff_url)
    response = requests.get(ff_url)
    response.raise_for_status()

    aln = Align.from_stockholm(response.text)
    return aln

def load_all_funfam_info(cath_version, *, cache_file=None, nocache=False):
    """
    Loads Funfam info from API
    """

    funfam_url = API_FUNFAM_BASE.format(cath_version=cath_version)

    if not cache_file:
        cache_file = f'cath.{cath_version}.funfam.json'

    # write the API json data to a local cache
    if not os.path.isfile(cache_file) or nocache:

        LOG.info(f"Fetching Funfam data from '{funfam_url}' ...")
        response = requests.get(funfam_url)
        response.raise_for_status()
        LOG.info("   ... done")

        LOG.info(f"Writing to cache file '{cache_file}'")
        with open(cache_file, 'wt') as fh:
            fh.write(response.text)

    # read the json
    with open(cache_file) as fh:
        LOG.info(f"Loading cached Funfam data from '{cache_file}' ...")
        all_funfam_data = json.load(fh)

    funfams = []
    for ff_data in all_funfam_data['data']:
        ff = models.FunfamInfo(**ff_data, cath_version=cath_version)
        funfams.extend([ff])

    return funfams


scripts/02.medium.py
#!/usr/bin/env python3
"""
Filter all funfams by:
    - high dops (>70)
    - at least 5 sequences

Then include funfams with the following characteristics:
    - lots of sequences
    - not many sequences
    - with structural reps
    - without structural reps

Within all those groups, select:
    - variety of unique folds

"""

# core
import json
import logging
import os
import os.path as path

# pip
import pandas as pd

# local
from cathbaker import loaders, models 

# params
MIN_DOPS_SCORE = 70
MIN_SEQUENCES = 5
LOW_SEQUENCE_COUNT = 50
HIGH_SEQUENCE_COUNT = 5000

# filenames
DATASET_DIR = path.relpath(path.join(path.dirname(__file__), '..', 'dataset', '02.medium'))
FF_ALN_DIR = path.join(DATASET_DIR, 'alignments')
FF_ALN_DATAFILE = path.join(DATASET_DIR, '02.medium.{category}.tsv')

# logging
logging.basicConfig(level='INFO', format='%(message)s')
LOG = logging.getLogger(__name__)

def log_title(msg, *args):
    """Formats 'title' log entries"""
    LOG.info('')
    LOG.info(msg, *args)
    LOG.info('-' * len(msg))

def log_kv(key, value):
    """Formats key/value log entries"""
    LOG.info('{:<60s} {}'.format(key, value))

def main():
    """Generates the benchmark datasets"""

    if not path.exists(FF_ALN_DIR):
        LOG.info(f"Creating alignment directory '{FF_ALN_DIR}'")
        os.makedirs(FF_ALN_DIR)

    log_title('LOADING FUNFAM DATA')
    all_funfams = loaders.load_all_funfam_info(cath_version='v4_2_0')

    # for ff in all_funfams[:3]:
    #     LOG.info("ff: %s", ff)

    # unique superfamily ids
    all_sfam_ids = {ff.superfamily_id for ff in all_funfams}
    all_top_ids = {ff.topology_id for ff in all_funfams}
    LOG.info('Found %s funfams from %s unique superfamilies and %s unique folds', 
        len(all_funfams), len(all_sfam_ids), len(all_top_ids))

    # apply global filters
    log_title('APPLYING GLOBAL FILTERS')
    filtered_funfams = [ff for ff in all_funfams if ff.seed_dops_score >= MIN_DOPS_SCORE and ff.num_members_in_seed_aln > MIN_SEQUENCES]
    log_kv(f'High sequence diversity (seed_dops_score >= {MIN_DOPS_SCORE}):', 
        f'{len(filtered_funfams)} (from {len(all_funfams)})')

    # create categories (high seqs, low seqs, with structural reps, without structural reps)
    log_title('CREATING DATASET CATEGORIES')
    low_seq_ffs = [ff for ff in filtered_funfams if ff.num_members_in_seed_aln < LOW_SEQUENCE_COUNT]
    log_kv(f'Small Funfams (num_members < {LOW_SEQUENCE_COUNT}):',
        f'{len(low_seq_ffs)} (from {len(filtered_funfams)})')

    high_seq_ffs = [ff for ff in filtered_funfams if ff.num_members_in_seed_aln >= HIGH_SEQUENCE_COUNT]
    log_kv(f'Large Funfams (num_members >= {HIGH_SEQUENCE_COUNT}):', 
        f'{len(high_seq_ffs)} (from {len(filtered_funfams)})')

    with_str_ffs = [ff for ff in filtered_funfams if ff.rep_source_id == 'cath']
    log_kv(f'Funfams with structure (rep_source_id="cath"):',
        f'{len(with_str_ffs)} (from {len(filtered_funfams)})')

    without_str_ffs = [ff for ff in filtered_funfams if ff.rep_source_id == 'uniprot']
    log_kv(f'Funfams without structure (rep_source_id="uniprot"):',
        f'{len(without_str_ffs)} (from {len(filtered_funfams)})')

    log_title('WRITING FUNFAM ALIGNMENTS')

    found_topologies = set()

    df_all = None
    for dataset_title_value in [
            ['low sequences', 'low_seq', low_seq_ffs],
            ['high sequences', 'high_seq', high_seq_ffs],
            ['structure', 'with_str', with_str_ffs],
            ['no structure', 'without_str', without_str_ffs],
        ]:
        title, category, ffs = dataset_title_value

        log_title(f'Working on {len(ffs)} funfams with {title} ...')
        ff_by_topology = {}
        for ff in ffs:
            if ff.topology_id not in found_topologies:
                found_topologies.add(ff.topology_id)
                ff_by_topology[ff.topology_id] = ff
        topology_ffs = list(ff_by_topology.values())
        log_kv('Funfams with unique topologies:', 
            f'{len(topology_ffs)} (from {len(all_top_ids)})')

        df = retrieve_funfam_alignments(topology_ffs)

        # add 'category' column
        df['category'] = category

        # datafile = FF_ALN_DATAFILE.format(category=category)
        # LOG.info(f"Writing TSV datafile '{datafile}'")
        # df.to_csv(datafile, sep='\t', index=False)

        if df_all is None:
            df_all = df
        else:
            df_all = pd.concat([df_all, df], ignore_index=True)

    log_title('WRITING DATASETS')
    log_kv(f'Total number of funfams', f'{len(df_all.index)} (from {len(all_funfams)})' )
    datafile = FF_ALN_DATAFILE.format(category='all')
    LOG.info(f"Writing TSV datafile '{datafile}'")
    df_all.to_csv(datafile, sep='\t', index=False)

    LOG.info('DONE')


def retrieve_funfam_alignments(funfams, dirpath=FF_ALN_DIR):
    """
    Retrieves funfam alignments from API
    
    Returns:
        df (pandas.DataFrame): dataframe with a summary of aln data
    """

    colnames = ['ff_id', 'sfam_id', 'seq_count', 'dops_score', 'gap_per']
    rows = []
    for ff in funfams:
        aln = loaders.get_funfam_alignment(**(ff.__dict__))
        aln_meta = aln.get_meta_summary()

        ff_id = f'{ff.superfamily_id}-ff-{ff.funfam_number}'
        ff_fname = f'cath.{ff.cath_version}.{ff.superfamily_id}-ff-{ff.funfam_number}.seed.sto'

        if aln_meta.dops_score < MIN_DOPS_SCORE:
            LOG.warning(f'WARNING: {ff_id}: alignment has dops_score={aln_meta.dops_score} (API reported: {ff.seed_dops_score}) (SKIPPING)')
            continue

        if aln_meta.seq_count is not ff.num_members_in_seed_aln:
            LOG.warning(f'WARNING: {ff_id}: alignment has seq_count={aln_meta.seq_count} (API reported: {ff.num_members_in_seed_aln})')

        local_file = path.join(dirpath, ff_fname)
        LOG.info(f"Saving Funfam alignment {ff_id}")
        aln.write_sto(local_file)
        
        gap_per = 100 * aln.total_gap_positions / aln.total_positions

        row = [ff_id, ff.superfamily_id, aln_meta.seq_count, aln_meta.dops_score, gap_per]
        rows.extend([row])

    df = pd.DataFrame(data=rows, columns=colnames)

    return df


if __name__ == '__main__':
    main()